rita_exit/src/database/struct_tools.rs

```diff
@@ -1,89 +1,246 @@
 use althea_kernel_interface::ExitClient;
-use althea_types::ExitClientIdentity;
-use althea_types::Identity;
-use arrayvec::ArrayString;
-use exit_db::models;
-use exit_db::models::Client;
-use ipnetwork::IpNetwork;
-use rand::Rng;
-use std::collections::HashSet;
-use std::fmt::Write as _;
-use std::net::IpAddr;
-
-use crate::RitaExitError;
-
-pub fn to_identity(client: &Client) -> Result<Identity, Box<RitaExitError>> {
-    trace!("Converting client {:?}", client);
-    Ok(Identity {
-        mesh_ip: match client.mesh_ip.clone().parse() {
-            Ok(a) => a,
-            Err(e) => return Err(Box::new(e.into())),
-        },
-        eth_address: match client.eth_address.clone().parse() {
-            Ok(a) => a,
-            Err(e) => return Err(Box::new(e.into())),
-        },
-        wg_public_key: match client.wg_pubkey.clone().parse() {
-            Ok(a) => a,
-            Err(e) => return Err(Box::new(e.into())),
-        },
-        nickname: Some(ArrayString::<32>::from(&client.nickname).unwrap_or_default()),
-    })
+use althea_types::{Identity, WgKey};
+use ipnetwork::{IpNetwork, Ipv4Network};
+use std::collections::hash_map::DefaultHasher;
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+use std::fmt::Write;
+use std::hash::{Hash, Hasher};
+use std::net::{IpAddr, Ipv4Addr};
+use std::sync::{Arc, RwLock};
+
+use crate::{generate_iterative_client_subnet, RitaExitError, DEFAULT_CLIENT_SUBNET_SIZE};
+
+/// Wg exit port on client side
+pub const CLIENT_WG_PORT: u16 = 59999;
+
+/// Max number of time we try to generate a valid ip addr before returning an eror
+pub const MAX_IP_RETRIES: u8 = 10;
+
+lazy_static! {
+    /// Keep track of ip addrs assigned to clients and ensure collisions dont happen. In worst case
+    /// the exit restarts and loses all this data in which case those client they had collision may get new
+    /// ip addrs and would need to setup wg exit tunnel again
+    static ref IP_ASSIGNMENT_MAP: Arc<RwLock<IpAssignmentMap>> = Arc::new(RwLock::new(IpAssignmentMap::default()));
```

Comment:
it's a bit of a nit but I'd call this file 'in memory database' because effectively we're building an in memory database based on what we get from the smart contract. Also we want there to be a single lazy static for all the exit setup state so that different lazy statics remain internally consistent. Also this needs to be indexed for the test env.

Reply:
done

Reply:
done